pkg/scheduler/preemption/preemption.go

```diff
@@ -255,7 +258,38 @@ func restoreSnapshot(snapshot *cache.Snapshot, targets []*workload.Info) {
 	}
 }
 
-func fairPreemptions(wl *workload.Info, assignment flavorassigner.Assignment, snapshot *cache.Snapshot, resPerFlv resourcesPerFlavor, candidates []*workload.Info, allowBorrowingBelowPriority *int32) []*workload.Info {
+type fsStrategy func(preemptorNewShare, preempteeOldShare, preempteeNewShare int) bool
+
+// lessThanOrEqualToFinalShare implements Rule S2-a in https://sigs.k8s.io/kueue/keps/1714-fair-sharing#choosing-workloads-from-clusterqueues-for-preemption
+func lessThanOrEqualToFinalShare(preemptorNewShare, _, preempteeNewShare int) bool {
+	return preemptorNewShare <= preempteeNewShare
+}
+
+// lessThanInitialShare implements rule S2-b in https://sigs.k8s.io/kueue/keps/1714-fair-sharing#choosing-workloads-from-clusterqueues-for-preemption
+func lessThanInitialShare(preemptorNewShare, preempteeOldShare, _ int) bool {
+	return preemptorNewShare < preempteeOldShare
+}
+
+// parseStrategies converts an array of strategies into the functions to the used by the algorithm.
+// This function takes advantage of the properties of the preemption algorithm and the strategies.
+// The number of functions returned might not match the input slice.
+func parseStrategies(s []config.PreemptionStrategy) []fsStrategy {
+	result := []fsStrategy{lessThanOrEqualToFinalShare}
+	if len(s) == 0 {
+		return result
+	}
+	if s[0] == config.LessThanInitialShare {
+		result[0] = lessThanInitialShare
+		// This rule is a superset of the other rule, no need to check other strategies.
```

Comment:
Are we confident Rule B is superset of Rule A? 

I'm thinking it could be possible that `preempteeOldShare=preempteeNewShare` (for example, there are 2 resources with the same DRF, and the workload being preempted reduces DRF only for one resource, so the DRF for the entire CQ stays the same (because it is a max).

In that case applying Rule A could help, while Rule B does not help.

Reply:
Yeah, that's a possibility, although unlikely. I could make the 4 combinations work. But Let's see what we decide about the overall topic.

Reply:
Since we go for the "mathematical" approach, I would suggest to be precise, and don't make the assumption that Rule B is a superset of Rule A (since unlikely scenarios may happen).

Reply:
Removed this return in favor of validation.